CMakeLists.txt

```diff
@@ -793,7 +793,12 @@ endif()
 option(ENABLE_UNIX_SOCKETS "Define if you want Unix domain sockets support" ON)
 if(ENABLE_UNIX_SOCKETS)
   include(CheckStructHasMember)
-  check_struct_has_member("struct sockaddr_un" sun_path "sys/un.h" USE_UNIX_SOCKETS)
+  if(WIN32)
+    check_struct_has_member("struct sockaddr_un" sun_path
+                            "${CMAKE_CURRENT_SOURCE_DIR}/lib/unixsockets_win32.h" USE_UNIX_SOCKETS)
```

Comment:
CMake will print a summary in the end, like this:
```
-- Enabled features: SSL IPv6 libz brotli zstd AsynchDNS IDN Largefile SSPI alt-svc HSTS SPNEGO Kerberos NTLM HTTP2 unicode
-- Enabled protocols: DICT FILE FTP FTPS GOPHER GOPHERS HTTP HTTPS IMAP IMAPS LDAP LDAPS MQTT POP3 POP3S RTSP SCP SFTP SMB SMBS SMTP SMTPS TELNET TFTP
-- Enabled SSL backends: Schannel
```
The check of unixsockets should be done by cmake. So that we can see "unixsockets" in "Enabled features" before compiling.

Reply:
You move Unix Sockets detection to curl_setup.h, this means whether unix sockets enable of disable will be automatically determined, so that cmake option  "ENABLE_UNIX_SOCKETS" doesn't work on Windows. Users cannot disable unix sockets feature by seting ENABLE_UNIX_SOCKETS to OFF.

Reply:
Could this code work for detection in CMake?:
```cmake
  if(WIN32)
    check_struct_has_member("struct sockaddr_un" sun_path
      "${CMAKE_CURRENT_SOURCE_DIR}/lib/curl_setup.h" USE_UNIX_SOCKETS)
```

Reply:
Your patch could not work.
```
-- Performing Test USE_UNIX_SOCKETS
-- Performing Test USE_UNIX_SOCKETS - Failed
```
In CMakeError.log:
```
Performing C SOURCE FILE Test USE_UNIX_SOCKETS failed with the following output:
Change Dir: D:/curl/0build/CMakeFiles/CMakeTmp

Run Build Command(s):D:/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe cmTC_62208 && [1/2] Building C object CMakeFiles\cmTC_62208.dir\src.c.obj

FAILED: CMakeFiles/cmTC_62208.dir/src.c.obj 

"D:\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\bin\Hostx86\x86\cl.exe"  /nologo -DUSE_UNIX_SOCKETS -ID:\curl\curl\include /DWIN32 /D_WIN32 /DNGHTTP2_STATICLIB  /DWIN32 /D_WINDOWS  /Zi /Ob0 /Od /RTC1 -MDd    -D_WIN32_WINNT=0x0600 -D_WINSOCKAPI_= -DSECURITY_WIN32 /showIncludes /FoCMakeFiles\cmTC_62208.dir\src.c.obj /FdCMakeFiles\cmTC_62208.dir\ /FS -c src.c

D:\curl\curl\include\curl/curl.h(134): error C2061: syntax error: identifier 'curl_socket_t'
D:\curl\curl\include\curl/curl.h(134): error C2059: syntax error: ';'
D:\curl\curl\include\curl/curl.h(402): error C2146: syntax error: missing ')' before identifier 'curlfd'
D:\curl\curl\include\curl/curl.h(402): error C2081: 'curl_socket_t': name in formal parameter list illegal
D:\curl\curl\include\curl/curl.h(402): error C2061: syntax error: identifier 'curlfd'
D:\curl\curl\include\curl/curl.h(402): error C2059: syntax error: ';'
D:\curl\curl\include\curl/curl.h(402): error C2059: syntax error: ','
D:\curl\curl\include\curl/curl.h(403): error C2059: syntax error: ')'
D:\curl\curl\include\curl/curl.h(412): error C2079: 'addr' uses undefined struct 'sockaddr'
D:\curl\curl\include\curl/curl.h(416): error C2143: syntax error: missing ')' before '*'
D:\curl\curl\include\curl/curl.h(416): error C2143: syntax error: missing '{' before '*'
D:\curl\curl\include\curl/curl.h(416): error C2059: syntax error: ')'
D:\curl\curl\include\curl/curl.h(421): error C2146: syntax error: missing ')' before identifier 'item'
D:\curl\curl\include\curl/curl.h(421): error C2081: 'curl_socket_t': name in formal parameter list illegal
D:\curl\curl\include\curl/curl.h(421): error C2061: syntax error: identifier 'item'
D:\curl\curl\include\curl/curl.h(421): error C2059: syntax error: ';'
D:\curl\curl\include\curl/curl.h(421): error C2059: syntax error: ')'
D:\curl\curl\include\curl\multi.h(115): error C2061: syntax error: identifier 'curl_socket_t'
D:\curl\curl\include\curl\multi.h(118): error C2059: syntax error: '}'
D:\curl\curl\include\curl\multi.h(159): error C2143: syntax error: missing ')' before '*'
D:\curl\curl\include\curl\multi.h(159): error C2081: 'fd_set': name in formal parameter list illegal
D:\curl\curl\include\curl\multi.h(159): error C2143: syntax error: missing '{' before '*'
D:\curl\curl\include\curl\multi.h(160): error C2143: syntax error: missing ';' before '*'
D:\curl\curl\include\curl\multi.h(161): error C2143: syntax error: missing ';' before '*'
D:\curl\curl\include\curl\multi.h(162): error C2059: syntax error: 'type'
D:\curl\curl\include\curl\multi.h(162): error C2059: syntax error: ')'
D:\curl\curl\include\curl\multi.h(296): error C2146: syntax error: missing ')' before identifier 's'
D:\curl\curl\include\curl\multi.h(296): error C2081: 'curl_socket_t': name in formal parameter list illegal
D:\curl\curl\include\curl\multi.h(296): error C2061: syntax error: identifier 's'
D:\curl\curl\include\curl\multi.h(296): error C2059: syntax error: ';'
D:\curl\curl\include\curl\multi.h(296): error C2059: syntax error: ','
D:\curl\curl\include\curl\multi.h(300): error C2059: syntax error: ')'
D:\curl\curl\include\curl\multi.h(317): error C2146: syntax error: missing ')' before identifier 's'
D:\curl\curl\include\curl\multi.h(317): error C2081: 'curl_socket_t': name in formal parameter list illegal
D:\curl\curl\include\curl\multi.h(317): error C2061: syntax error: identifier 's'
D:\curl\curl\include\curl\multi.h(317): error C2059: syntax error: ';'
D:\curl\curl\include\curl\multi.h(317): error C2059: syntax error: ','
D:\curl\curl\include\curl\multi.h(318): error C2059: syntax error: ')'
D:\curl\curl\include\curl\multi.h(321): error C2146: syntax error: missing ')' before identifier 's'
D:\curl\curl\include\curl\multi.h(321): error C2081: 'curl_socket_t': name in formal parameter list illegal
D:\curl\curl\include\curl\multi.h(321): error C2061: syntax error: identifier 's'
D:\curl\curl\include\curl\multi.h(321): error C2059: syntax error: ';'
D:\curl\curl\include\curl\multi.h(321): error C2059: syntax error: ','
D:\curl\curl\include\curl\multi.h(323): error C2059: syntax error: ')'
D:\curl\curl\include\curl\multi.h(423): error C2146: syntax error: missing ')' before identifier 'sockfd'
D:\curl\curl\include\curl\multi.h(423): error C2081: 'curl_socket_t': name in formal parameter list illegal
D:\curl\curl\include\curl\multi.h(423): error C2061: syntax error: identifier 'sockfd'
D:\curl\curl\include\curl\multi.h(423): error C2059: syntax error: ';'
D:\curl\curl\include\curl\multi.h(423): error C2059: syntax error: ','
D:\curl\curl\include\curl\multi.h(423): error C2059: syntax error: ')'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(139): error C2371: 'fd_set': redefinition; different basic types
D:\curl\curl\include\curl\multi.h(161): note: see declaration of 'fd_set'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(141): error C2143: syntax error: missing ')' before '*'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(141): error C2081: 'fd_set': name in formal parameter list illegal
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(141): error C2143: syntax error: missing '{' before '*'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(141): error C2059: syntax error: ')'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(141): error C2059: syntax error: ';'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(2054): error C2143: syntax error: missing ')' before '*'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(2054): error C2081: 'fd_set': name in formal parameter list illegal
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(2054): error C2143: syntax error: missing '{' before '*'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(2055): error C2143: syntax error: missing ';' before '*'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(2056): error C2143: syntax error: missing ';' before '*'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(2057): error C2059: syntax error: 'type'
D:\Windows Kits\10\include\10.0.19041.0\um\winsock2.h(2058): error C2059: syntax error: ')'
D:\curl\curl\lib\curl_setup.h(818): warning C4005: 'USE_UNIX_SOCKETS': macro redefinition
src.c: note: see previous definition of 'USE_UNIX_SOCKETS'
ninja: build stopped: subcommand failed.



Source file was:

#include <D:/curl/curl/lib/curl_setup.h>

int main()
{
  (void)sizeof(((struct sockaddr_un *)0)->sun_path);
  return 0;
}
```
curl_setup.h includes many other header files, it is not proper to be used for a check.

Reply:
Have your test included this line:
```
set(CMAKE_REQUIRED_INCLUDES "${CMAKE_CURRENT_SOURCE_DIR}/include")
```